icu4c/source/i18n/unitsdata.cpp

```diff
@@ -0,0 +1,194 @@
+// © 2020 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+
+#include "unicode/utypes.h"
+
+#if !UCONFIG_NO_FORMATTING
+
+#include <utility>
+
+#include "cstring.h"
+#include "resource.h"
+#include "unitsdata.h"
+#include "uresimp.h"
+
+U_NAMESPACE_BEGIN
+
+namespace {
+
+/**
+ * A ResourceSink that collects conversion rate information.
+ *
+ * This class is for use by ures_getAllItemsWithFallback. Example code for
+ * collecting conversion info for "mile" and "foot" into conversionInfoOutput:
+ *
+ *     UErrorCode status = U_ZERO_ERROR;
+ *     ures_getByKey(unitsBundle, "convertUnits", &fillIn, &status);
+ *     MaybeStackVector<ConversionRateInfo> conversionInfoOutput;
+ *     ConversionRateDataSink convertSink(conversionInfoOutput);
+ *     ures_getAllItemsWithFallback(fillIn, "mile", convertSink, status);
+ *     ures_getAllItemsWithFallback(fillIn, "foot", convertSink, status);
+ */
+class ConversionRateDataSink : public ResourceSink {
+  public:
+    /**
+     * Constructor.
+     * @param out The vector to which ConversionRateInfo instances are to be
+     * added.
+     */
+    explicit ConversionRateDataSink(MaybeStackVector<ConversionRateInfo> &out) : outVector(out) {}
+
+    /**
+     * Adds the conversion rate information found in value to the output vector.
+     *
+     * Each call to put() collects a ConversionRateInfo instance for the
+     * specified source unit identifier into the vector passed to the
+     * constructor, but only if an identical instance isn't already present.
+     *
+     * @param source The source unit identifier.
+     * @param value A resource containing conversion rate info (the base unit
+     * and factor, and possibly an offset).
+     * @param noFallback Ignored.
+     * @param status The standard ICU error code output parameter.
+     */
+    void put(const char *source, ResourceValue &value, UBool /*noFallback*/, UErrorCode &status) {
+        if (U_FAILURE(status)) return;
+        ResourceTable conversionRateTable = value.getTable(status);
+        if (U_FAILURE(status)) return;
+
+        // Collect base unit, factor and offset from the resource.
+        int32_t lenSource = uprv_strlen(source);
+        const UChar *baseUnit = NULL, *factor = NULL, *offset = NULL;
```

Comment:
Better ICU4C style:

1. Make these UnicodeStrings and make them default to bogus: there's something like `Utility.makeBogusString()` to help you do that.
2. Eliminate the length variables.
3. When getting the value in the loop, get it as a fixed-length UnicodeString and set it into the variable.
4. You can check whether any of the strings are still bogus, instead of checking for nullptr.

Reply:
This is related to Hugo pull request: #31

Reply:
I think https://github.com/sffc/icu/pull/31/commits/2deeb28f3c51aa16e351e6c70ea547cb5cef6679 fixes this: using UnicodeString, ResourceValue::getUnicodeString, and appendInvariantChars() that takes a UnicodeString.

Thanks!